### PR TITLE
Feature/unr 4198 add async load filter

### PR DESCRIPTION
**Pull request by m-samiec**
#### Description
Re-added async loading, via a new async load filter.

#### Tests
https://github.com/spatialos/UnrealGDKTestGyms/pull/308

#### Primary reviewers
@jacquese @samiwh 
Based on work previously done by @improbable-valentyn and @ImprobableNic

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	UObject* ClassObject = FindObject<UClass>(nullptr, *ClassPath, false);
+	if (ClassObject == nullptr)
+	{
+		return true;
+	}
+
+	FString PackagePath = GetPackagePath(ClassPath);
+	FName PackagePathName = *PackagePath;
+
+	// UNR-3320 The following test checks if the package is currently being processed in the async loading thread.
```
Comment (jacquese) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:46]: are these tickets duplicates?

**Review by jacquese — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		AsyncLoadingPackages.Remove(PackageName);
+		return;
+	}
+
+	LoadedPackages.Add(PackageName);
+}
+
+void AsyncPackageLoadFilter::ProcessActorsFromAsyncLoading()
+{
+	static_assert(TContainerTraits<decltype(LoadedPackages)>::MoveWillEmptyContainer, "Moving the set won't empty it");
```
[the remark this answers is not in the document]
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:107]: This also seems like a really specific test that I'd never expect to fail for a TSet. Is there something we've been bitten by before?

[the remark this answers is not in the document]
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:107]: Can we not just iterate over the loaded packages then clear it? Unless there's some kind of threading thing going on don't we need to synchronise it.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	TSet<FName> PackagesToProcess = MoveTemp(LoadedPackages);
+
+	for (const auto& PackageName : PackagesToProcess)
+	{
+		TArray<Worker_EntityId> Entities;
+		if (!AsyncLoadingPackages.RemoveAndCopyValue(PackageName, Entities))
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:113]: There's a lot of array copying here. Would it be better to just get an iterator and then remove it at the end?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+				   TEXT("USpatialReceiver::OnAsyncPackageLoaded: Package loaded but no entry in AsyncLoadingPackages. Package: %s"),
+				   *PackageName.ToString());
+			continue;
+		}
+
+		for (Worker_EntityId Entity : Entities)
+		{
+			if (IsEntityWaitingForAsyncLoad(Entity))
+			{
+				UE_LOG(LogAsyncPackageLoadFilter, Log, TEXT("Finished async loading package %s for entity %lld."), *PackageName.ToString(),
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:125]: Nit - Can we put the copyable message first then do the dynamic stuff at the end. This mostly does that already.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:107]: That's to make sure we do not have any further re-entrancy issues. Maybe processing these packages will keep pushing stuff into the LoadedPackages Set? Don't know, don't care.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+DEFINE_LOG_CATEGORY(LogAsyncPackageLoadFilter);
+
+namespace SpatialGDK
+{
+AsyncPackageLoadFilter::AsyncPackageLoadFilter(USpatialNetDriver* InNetDriver)
+	: NetDriver(InNetDriver)
+{
+}
+
+bool AsyncPackageLoadFilter::ProcessAndCheckEntityFilter(Worker_EntityId EntityId, const FString& ClassPath)
```
[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:19]: Oh I did spend at least 10 mins trying to come up with a better name here. IsLoadedElseLoad could work. Will have a think.

My reasoning behind returning true was that this is a filter, and what it does is filter things. Did it filter it? Yes it did, return true. Did it not filter it and thus is no stopped, no? return false.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:46]: This is from the old async loading code, will change the comment to indicate the ticket refers to the context, not the open issue.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp (hunk 4 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:125]: I don't follow what you mean

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:107]: This logic seems fine to me - prevents potential misuse in future - will leave as is unless anyone objects.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:19]: on filters returning true- my mental model for filters was true means it passed the filter, rather than it was filtered. We could change everything to match the idea that true means filtered out, however that would be a bigger refactor, and I would rather for consistency reasons we went with what's already there for now if we don't do the refactor. "true means complete" is also documented in the subview documentation here https://improbableio.atlassian.net/wiki/spaces/GBU/pages/1842086036/Subviews

WDYT? Happy to revisit the boolean if it doesn't make sense

[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:19]: Precedence is more than enough. Was just explaining my mental model. Will update 👍

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp (hunk 4 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:125]: I believe Sami is referring to this:
```suggestion
				UE_LOG(LogAsyncPackageLoadFilter, Log, TEXT("Finished async loading package for entity. Package: %s, entity: %lld."), *PackageName.ToString(),
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	// In practice, these tests are not enough to prevent using objects too early (symptom is RF_NeedPostLoad being set, and crash when
+	// using them later). GetAsyncLoadPercentage will actually look through the async loading thread's UAsyncPackage maps to see if there
+	// are any entries. See UNR-3320 for more context.
+	// TODO : UNR-3374 This looks like an expensive check, but it does the job. We should investigate further
+	// what is the issue with the other flags and why they do not give us reliable information.
+
+	float Percentage = GetAsyncLoadPercentage(PackagePathName);
+	if (Percentage != -1.0f)
+	{
+		UE_LOG(LogAsyncPackageLoadFilter, Warning, TEXT("Class %s package is registered in async loading thread."), *ClassPath)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:59]:
```suggestion
		UE_LOG(LogAsyncPackageLoadFilter, Warning, TEXT("Class package is registered in async loading thread. Class path: %s"), *ClassPath)
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+{
+	FString PackagePath = GetPackagePath(ClassPath);
+	FName PackagePathName = *PackagePath;
+
+	bool bAlreadyLoading = AsyncLoadingPackages.Contains(PackagePathName);
+
+	EntitiesWaitingForAsyncLoad.Emplace(EntityId);
+	AsyncLoadingPackages.FindOrAdd(PackagePathName).Add(EntityId);
+
+	UE_LOG(LogAsyncPackageLoadFilter, Log, TEXT("Async loading package %s for entity %lld. Already loading: %s"), *PackagePath, EntityId,
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:85]:
```suggestion
	UE_LOG(LogAsyncPackageLoadFilter, Log, TEXT("Async loading package for entity. Package: %s, entity: %lld, already loading: %s"), *PackagePath, EntityId,
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+		   bAlreadyLoading ? TEXT("true") : TEXT("false"));
+	if (!bAlreadyLoading)
+	{
+		LoadPackageAsync(PackagePath, FLoadPackageAsyncDelegate::CreateRaw(this, &AsyncPackageLoadFilter::OnAsyncPackageLoaded));
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:89]: Is it possible for the class to finish loading after the filter object got destroyed? E.g. after a client travel.

File: SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/EntityFactory/EntityFactoryTest.cpp [hunk 9]
```diff
@@ -39,7 +39,7 @@ ENTITYFACTORY_TEST(GIVEN_an_actor_WHEN_creating_skeleton_entity_components_THEN_
 	TestTrue("Has actor auth tag component",
 			 ComponentDatas.ContainsByPredicate(ContainsComponent{ SpatialConstants::ACTOR_AUTH_TAG_COMPONENT_ID }));
 	TestTrue("Has actor non-auth tag component",
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKTests/SpatialGDK/Utils/EntityFactory/EntityFactoryTest.cpp:41]:
```suggestion
	TestTrue("Has actor tag component",
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+		for (Worker_EntityId_Key Entity : *Entities)
+		{
+			if (IsEntityWaitingForAsyncLoad(Entity))
+			{
+				UE_LOG(LogAsyncPackageLoadFilter, Log, TEXT("Finished async loading package %s for entity %lld."), *PackageName.ToString(),
+					   Entity);
+
+				check(EntitiesWaitingForAsyncLoad.Find(Entity) != nullptr);
+				EntitiesWaitingForAsyncLoad.Remove(Entity);
+				NetDriver->Connection->GetCoordinator().RefreshEntityCompleteness(Entity);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:130]: Would it help to remove the dependency on the net driver and worker connection by passing a delegate that we execute here instead?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+			}
+		}
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/AsyncPackageLoadFilter.cpp:132]:
```suggestion
		}

		AsyncLoadingPackages.Remove(PackageName);
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 12]
```diff
@@ -124,8 +125,6 @@ bool USpatialNetDriver::InitBase(bool bInitAsClient, FNetworkNotify* InNotify, c
 
 	FCoreUObjectDelegates::PostLoadMapWithWorld.AddUObject(this, &USpatialNetDriver::OnMapLoaded);
 
-	FWorldDelegates::LevelAddedToWorld.AddUObject(this, &USpatialNetDriver::OnLevelAddedToWorld);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:127]: Driveby removal of duplicate add that was done in base class

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 13]
```diff
@@ -410,7 +408,10 @@ void USpatialNetDriver::CreateAndInitializeCoreClasses()
 
 	if (GetDefault<USpatialGDKSettings>()->bAsyncLoadNewClassesOnEntityCheckout)
 	{
-		AsyncPackageLoadFilter = MakeUnique<SpatialGDK::AsyncPackageLoadFilter>(this);
+		AsyncPackageLoadFilter = NewObject<UAsyncPackageLoadFilter>();
+		AsyncPackageLoadFilter->Init(this);
+		AsyncPackageLoadFilter->GetOnPackageLoadedForEntityDelegate().AddDynamic(this,
```
Comment (jacquese) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:413]: can we pass the function reference into init instead?

**Review by jacquese — APPROVED** [review]
lgtm

File: SpatialGDK/Source/SpatialGDK/Public/Interop/AsyncPackageLoadFilter.h [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+	bool NeedToLoadClass(const FString& ClassPath);
+	bool IsEntityWaitingForAsyncLoad(Worker_EntityId Entity);
+	void StartAsyncLoadingClass(Worker_EntityId EntityId, const FString& ClassPath);
+
+	FString GetPackagePath(const FString& ClassPath);
+	void OnAsyncPackageLoaded(const FName& PackageName, UPackage* Package, EAsyncLoadingResult::Type Result);
+
+	FOnPackageLoadedForEntity OnPackageLoadedForEntity;
+
+	USpatialNetDriver* NetDriver;
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/AsyncPackageLoadFilter.h:40]: Not needed anymore.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 13 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:413]: I made the delegate a multicast so other things could listen to this event, however if we don't think that functionality is worthwhile, I can switch it to a single delegate and set via a single function.

**Review by improbable-valentyn — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 13 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:413]: I would slightly prefer the single delegate since it makes the code cleaner, can also use the simple delegate instead of dynamic so we can pass `Worker_EntityId` instead of `int64`. Should be safe to do so because the delegate can only be called when the filter object is valid.
